docs/Controllers.md

```diff
@@ -0,0 +1,87 @@
+
+Later API versions might use URNs which have not yet been defined and so Controllers MUST be tolerant to these.
+
+#### Error Codes & Responses
+The NMOS APIs use HTTP status codes to indicate success, failure and other cases to Controllers as per [RFC 7231](https://tools.ietf.org/html/rfc7231) and related standards.
+
+As explicit handling of every possible HTTP response code is not expected, Controllers MUST implement generic handling for ranges of response codes (1xx, 2xx, 3xx, 4xx and 5xx).
+However, where the RAML specification of an API specifies explicit response codes the Controller SHOULD handle these cases explicitly.
+
+For Controllers performing GET and HEAD requests, using these methods SHOULD correctly handle a 301 (Moved Permanently) response.
+
+When a 301 is supported, the Controller MUST follow the redirect in order to retrieve the required response payload.
+
+If a Controller receives a HTTP 500 response code from the API, a failure has occurred.
+The Controller SHOULD display the content of the response�s error field to the user if possible, and indicate that the Resource may be in an unknown state.
+The Controller SHOULD also refresh the endpoints of the relevant Resources to ensure the Controller is accurately reflecting the current state of the API.
+
+## Registry Service Discovery	
+
+In order to locate the Registry, the Controller SHOULD support all of the following: 
+* Unicast DNS Service Discovery (DNS-SD).
+* mDNS.
+* Direct configuration with the location of a preferred Registry.
+
+The Controller SHOULD offer unicast DNS-SD as the default mechanism. 
+
+Controllers SHOULD observe and interpret all of the TXT records returned with the DNS Service Discovery responses according to the requirements for Query API Clients specified in the [API Paths section of the APIs document](APIs.md#api-paths) in this specification.
+
+## Query API
+The Controller MUST be capable of using the Registry's IS-04 Query API to discover any registered resource, including Node, Device,  Source, Flow, Sender, and Receiver,
+as described in the [APIs section](README.md#apis) of this specification.
+
+The Controller MUST use the Registry�s IS-04 Query API either via the REST API or by requesting WebSocket subscriptions.
+
+When using the Query API, basic queries SHOULD be used (and advanced query language where available) to cut down on the volume of resources returned to the Controller.	
+
+Controllers SHOULD adhere to the version downgrade requirements for Query API Clients specified in the [Requirements for Query API Clients section of the Upgrade Path document](Upgrade%20Path.md#requirements-for-query-api-clients) in this specification.
+
+## Pagination
+In large systems, API resources rely upon pagination in order to return high volumes of data.
+This can be both hard to keep track of for a client, and require a large number of requests in order to scan the entire data set.
```

Comment:
This is sourced from the [AMWA Wiki](https://github.com/AMWA-TV/nmos/wiki/Generic-API-Client#websockets--subscriptions).  Is the problem the wording or the intent?

Reply:
Understood, although the Wiki recommendations do not appear to conflict with IS-04, just express preference for subscription over REST API. @peterbrightwell do you know the provenance of the 'advice' regarding a preference for subscriptions?

Also, while noting the lower status of the Wiki, the Wiki's NMOS Glossary *is* referenced by [IS-10](https://github.com/AMWA-TV/is-10/blob/019e0e3b4181addabbc2e526c79f44d99b09e20d/docs/Definitions.md) and [BCP-003-03](https://github.com/AMWA-TV/bcp-003-03/blob/7f72f432e65d578230adb800fc16d9ca1043795e/docs/Certificate%20Provisioning.md) giving it a normative air. However, also worth noting that this now links [to a page](https://github.com/AMWA-TV/nmos/wiki/Glossary), which in turn contains a [broken link](https://specs.amwa.tv/nmos/branches/main/docs/4.0._Glossary.html).

Reply:
The advice on the wiki was based on common sense at the time, as paging through a lot of results to find a small change for updating could be inefficient.  Maybe there are other ways for finding incremental updates that would be helpful but that would be require new investigations.

The Wiki shouldn't ever be referenced by the specs (so any links should be changed).  The Glossary is now at <https://specs.amwa.tv/nmos/branches/main/docs/Glossary.html>.

Reply:
> Maybe there are other ways for finding incremental updates that would be helpful but that would be require new investigations.

😱 I don't think it requires new investigations. The whole paging mechanism is designed to make that easy!

1. The Query API defaults to `paging.order=update` on requests.
2. The `Link` response header provides the URL to request the next lot of updates (`rel="next"`).
3. That URL will use the `paging.since` query parameter with the response `X-Paging-Until` header value.

Reply:
Resolved.